src/shared/s3_manager.py

```diff
@@ -238,3 +271,29 @@ def write_local_metadata(self, metadata: dict | None = None, meta_type: str | No
             metadata=metadata,
             meta_type=meta_type,
         )
+
+    # Lockfile management
+    def request_or_validate_lock(self, lock_id: str | None = None):
+        try:
+            lock = functions.get_s3_json_as_dict(bucket=self.s3_bucket_name, key=self.lockfile)
+            if lock_id != lock["id"]:
+                raise errors.AggregatorStudyProcessingError
+
+        except botocore.exceptions.ClientError:
+            # if the lockfile doesn't exist, we'll make one
+            lock_id = str(uuid.uuid4())
+            lock = {"id": lock_id, "uploaded_at": datetime.datetime.now(datetime.UTC)}
+            self.s3_client.put_object(
```

Comment:
If you set [If-None-Match: *](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) as a header here, I think you avoid some race condition issues.